CRM/Utils/PDF/Utils.php

```diff
@@ -215,8 +215,12 @@ public static function _html2pdf_dompdf($paper_size, $orientation, $html, $outpu
   public static function _html2pdf_wkhtmltopdf($paper_size, $orientation, $margins, $html, $output, $fileName) {
     $config = CRM_Core_Config::singleton();
     // if the path doesn't exist fall back on the current backup which is DOMPDF.
-    if (!file_exists($config->wkhtmltopdfPath)) {
-      return self::_html2pdf_dompdf($paper_size, $orientation, $html, $output, $fileName);
+    if (!empty($config->wkhtmltopdfPath)) {
+      $pieces = explode(' ',$config->wkhtmltopdfPath);
+      $path = $pieces[0];
+      if (!file_exists($path) || !is_executable($path)) {
+        return self::_html2pdf_dompdf($paper_size, $orientation, $html, $output, $fileName);
```

Comment:
Rather than fallback to domPDF which actually produces very different PDF output. Would prefer that an error be raised, error logged and the operation cancelled. Some manual action is required to fix the issue.

Continuing the operation and sending a weird looking PDF is just as bad as not sending a PDF at all.

Reply:
kind of the point of the status check was that that was how the 'error' was being raised

Reply:
then why do this? return self::_html2pdf_dompdf

Reply:
The status check is incorrect, returns false positive - so the value of status check is "kind of" negated.

Reply:
@agileware-justin the reason to do that is because dompdf is the default fall back if wkhtmltopdf setting is not set https://github.com/civicrm/civicrm-core/blob/master/CRM/Utils/PDF/Utils.php#L104 so in this case we are saying that whilst it is set it can't be found so we are reverting back to what would be the default if the setting was just blank. 

My point was that the status check is how the error about if wkhtmltopdf can be run is alerted rather than in each individual potential PDF generation process. My opinion that is the more correct place.

I can accept that the status check / this code is giving some false positives but they should be fixed

Reply:
So Utils.php#L104 

Already implements the fall back if wkhtml is undefined, that's OK.

```
 if (CRM_Core_Config::singleton()->wkhtmltopdfPath) {
      return self::_html2pdf_wkhtmltopdf($paper_size, $orientation, $margins, $html, $output, $fileName);
    }
    else {
      return self::_html2pdf_dompdf($paper_size, $orientation, $html, $output, $fileName);
    }
```

Then in _html2pdf_wkhtmltopdf - I just don't understand why these checks are being performed again. Another check of the configuration, again check the paths and again use a fall back. This code seems redundant to me.

```
  if (!empty($config->wkhtmltopdfPath)) {
      $pieces = explode(' ',$config->wkhtmltopdfPath);
      $path = $pieces[0];
      if (!file_exists($path) || !is_executable($path)) {
        return self::_html2pdf_dompdf($paper_size, $orientation, $html, $output, $fileName);
```

Reply:
@seamuslee001 just a FYI you missed the opportunity to discuss this one at the virtual meet up yesterday :smile:

Reply:
@jusfreeman so the reason for the additional check here is for the following scenario

You move from one server to another (mysql dump and load in process etc) However on your new host wkhtmltopdf doesn't exist or doesn't exist in the same location.  However `wkhtmltopdfPath` is still set in civicrm_settings table

this if `if (CRM_Core_Config::singleton()->wkhtmltopdfPath) {` only checks to see that the CiviCRM setting is set. It does not check that the path actually is valid.

the new checks are saying ok we have the setting set but lets try and see if that path actually exists on the server before trying to run it because we ran into a hard fail with a client site where they were using the attach PDF receipt to a contribution receipt email. 

If we find that it isn't there or perhaps it can't be accessed by the www user then we are saying ok well lets not run through this function and use the backup function as we can't be sure that wkhtmltopdf is workable for us and we don't want to cause hard fails later on. 

The status check is there to prompt the users to ask for assistance because as you say they will likely be producing different content and the check in the utils is there to prevent any actual hard fails in between the point at which in my example the server migration occurs and either wkhtmltopdf is reinstalled or the setting is blanked out.